iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/DataNodeServerCommandLine.java

```diff
@@ -135,57 +163,60 @@ private void doRemoveDataNode(String[] args)
     }
 
     // Below supports multiple datanode deletion, split by ',', and is reserved for extension
-    try {
-      List<TEndPoint> endPoints = NodeUrlUtils.parseTEndPointUrls(args);
-      try (ConfigNodeClient client =
-          ConfigNodeClientManager.getInstance().borrowClient(ConfigNodeInfo.CONFIG_REGION_ID)) {
-        dataNodeLocations =
-            client.getDataNodeConfiguration(-1).getDataNodeConfigurationMap().values().stream()
-                .map(TDataNodeConfiguration::getLocation)
-                .filter(location -> endPoints.contains(location.getClientRpcEndPoint()))
-                .collect(Collectors.toList());
-      } catch (TException | ClientManagerException e) {
-        LOGGER.error("Get data node locations failed", e);
-      }
-    } catch (BadNodeUrlException e) {
-      try (ConfigNodeClient client =
-          ConfigNodeClientManager.getInstance().borrowClient(ConfigNodeInfo.CONFIG_REGION_ID)) {
-        for (String id : args.split(",")) {
-          if (!isNumeric(id)) {
-            LOGGER.warn("Incorrect id format {}, skipped...", id);
-            continue;
-          }
-          List<TDataNodeLocation> nodeLocationResult =
-              client
-                  .getDataNodeConfiguration(Integer.parseInt(id))
-                  .getDataNodeConfigurationMap()
-                  .values()
-                  .stream()
-                  .map(TDataNodeConfiguration::getLocation)
-                  .collect(Collectors.toList());
-          if (nodeLocationResult.isEmpty()) {
-            LOGGER.warn("DataNode {} is not in cluster, skipped...", id);
-            continue;
-          }
-          if (!dataNodeLocations.contains(nodeLocationResult.get(0))) {
-            dataNodeLocations.add(nodeLocationResult.get(0));
-          }
-        }
-      } catch (TException | ClientManagerException e1) {
-        LOGGER.error("Get data node locations failed", e);
-      }
+    List<NodeCoordinate> nodeCoordinates = parseCoordinates(args);
+    try (ConfigNodeClient client =
+        configNodeClientManager.borrowClient(ConfigNodeInfo.CONFIG_REGION_ID)) {
+      dataNodeLocations =
+          client.getDataNodeConfiguration(-1).getDataNodeConfigurationMap().values().stream()
+              .map(TDataNodeConfiguration::getLocation)
+              .filter(
+                  location ->
+                      nodeCoordinates.stream()
+                          .anyMatch(nodeCoordinate -> nodeCoordinate.matches(location)))
+              .collect(Collectors.toList());
+    } catch (TException | ClientManagerException e) {
+      LOGGER.error("Get data node locations failed", e);
     }
+
     return dataNodeLocations;
   }
+
+  protected List<NodeCoordinate> parseCoordinates(String coordinatesString) {
+    // Multiple nodeIds are separated by ","
+    String[] nodeIdStrings = coordinatesString.split(",");
+    List<NodeCoordinate> nodeIdCoordinates = new ArrayList<>(nodeIdStrings.length);
```

Comment:
1. can we use set now?
2. if it's empty. maybe we can do something?

Reply:
If it was completely empty we'd get an error earlier as the number of arguments wouldn't be right.
I could use a Set, however I used what I saw was used elsewhere. Also not quite sure what this sort of "optimization" would help with. The only thing I see would be if I added the raw ids in there, the way elements are added, the set would still contain duplicates. ArrayLists with fixed sizes are implemented more memory and cpu-efficient. So why should I change it? If the "de-duplicating" would be it, I would need to implement this differently. The benefit would be minimal however, as the way the filter is implemented it would simply match the "andMatch" of the first and still we'd get a de-duplicated list out in "dataNodeLocations".